File: UTILS - Logging/logger.js
// logger.js
// Simple logging utility for adding, reading, editing, and deleting log entries in a text file.
// All log entries are stored in 'log.txt' in the same directory as this script.

const fs = require('fs');
const LOG_FILE = 'log.txt'; // Name of the log file

/**
 * Add a new log entry to the log file.
 * @param {string} entry - The string to be added as a log entry.
 */
function addLog(entry) {
    fs.appendFileSync(LOG_FILE, entry + '\n', 'utf8'); // Append entry with newline
}

/**
 * Read all log entries from the log file.
 * @returns {string[]} Array of log entry strings. Returns an empty array if file does not exist.
 */
function readLogs() {
    if (!fs.existsSync(LOG_FILE)) return []; // Return empty array if log file doesn't exist
    return fs.readFileSync(LOG_FILE, 'utf8').split('\n').filter(Boolean); // Split by newline and remove empty lines
}

/**
 * Edit a specific log entry by line number (0-based).
 * @param {number} lineNumber - Index of the log entry to edit (0-based).
 * @param {string} newEntry - New string to replace the old log entry.
 * @returns {boolean} True if edit was successful, False otherwise.
 */
function editLog(lineNumber, newEntry) {
    let logs = readLogs();
    if (lineNumber >= 0 && lineNumber < logs.length) {
        logs[lineNumber] = newEntry; // Replace the specified line
        fs.writeFileSync(LOG_FILE, logs.join('\n') + '\n', 'utf8'); // Write all lines back to the file
        return true;
    }
    return false; // Return false if lineNumber is out of range
}

/**
 * Delete a specific log entry by line number (0-based).
 * @param {number} lineNumber - Index of the log entry to delete (0-based).
 * @returns {boolean} True if deletion was successful, False otherwise.
 */
function deleteLog(lineNumber) {
    let logs = readLogs();
    if (lineNumber >= 0 && lineNumber < logs.length) {
        logs.splice(lineNumber, 1); // Remove the specified line
        fs.writeFileSync(LOG_FILE, logs.join('\n') + '\n', 'utf8'); // Write remaining lines back to the file
        return true;
    }
    return false; // Return false if lineNumber is out of range
}

// Example usage (uncomment to use):
// addLog('This is a new log entry');
// console.log(readLogs());
// editLog(0, 'This is an edited log entry');
// deleteLog(0);

// Export functions for use in other files (if needed)
module.exports = { addLog, readLogs, editLog, deleteLog };

if (require.main === module) {
    // Menu-driven CLI for logging operations
    const readlineSync = require('readline-sync');
    while (true) {
        console.log('\n--- Log Menu ---');
        console.log('1. Add log entry');
        console.log('2. Read all log entries');
        console.log('3. Edit a log entry');
        console.log('4. Delete a log entry');
        console.log('5. Exit');
        const choice = readlineSync.question('Choose an option (1-5): ');
        if (choice === '1') {
            const entry = readlineSync.question('Enter a log entry: ');
            addLog(entry);
            console.log('Log entry added.');
        } else if (choice === '2') {
            const logs = readLogs();
            if (logs.length) {
                console.log('\nAll log entries:');
                logs.forEach((log, i) => console.log(`${i}: ${log}`));
            } else {
                console.log('No log entries found.');
            }
        } else if (choice === '3') {
            const logs = readLogs();
            if (!logs.length) {
                console.log('No log entries to edit.');
                continue;
            }
            logs.forEach((log, i) => console.log(`${i}: ${log}`));
            const idx = parseInt(readlineSync.question('Enter the line number to edit: '), 10);
            if (isNaN(idx) || idx < 0 || idx >= logs.length) {
                console.log('Invalid line number.');
                continue;
            }
            const newEntry = readlineSync.question('Enter the new log entry: ');
            if (editLog(idx, newEntry)) {
                console.log('Log entry edited.');
            } else {
                console.log('Failed to edit log entry.');
            }
        } else if (choice === '4') {
            const logs = readLogs();
            if (!logs.length) {
                console.log('No log entries to delete.');
                continue;
            }
            logs.forEach((log, i) => console.log(`${i}: ${log}`));
            const idx = parseInt(readlineSync.question('Enter the line number to delete: '), 10);
            if (isNaN(idx) || idx < 0 || idx >= logs.length) {
                console.log('Invalid line number.');
                continue;
            }
            if (deleteLog(idx)) {
                console.log('Log entry deleted.');
            } else {
                console.log('Failed to delete log entry.');
            }
        } else if (choice === '5') {
            console.log('Exiting log program.');
            break;
        } else {
            console.log('Invalid choice. Please select 1-5.');
        }
    }
} 

File: UTILS - Logging/logger.py
# logger.py
# Simple logging utility for adding, reading, editing, and deleting log entries in a text file.
# All log entries are stored in 'log.txt' in the same directory as this script.

import os

LOG_FILE = 'log.txt'  # Name of the log file

def add_log(entry):
    """
    Add a new log entry to the log file.
    :param entry: String to be added as a log entry.
    """
    with open(LOG_FILE, 'a', encoding='utf-8') as f:
        f.write(entry + '\n')  # Write the entry followed by a newline


def read_logs():
    """
    Read all log entries from the log file.
    :return: List of log entry strings. Returns an empty list if file does not exist.
    """
    if not os.path.exists(LOG_FILE):
        return []  # Return empty list if log file doesn't exist
    with open(LOG_FILE, 'r', encoding='utf-8') as f:
        return f.readlines()  # Return all lines as a list


def edit_log(line_number, new_entry):
    """
    Edit a specific log entry by line number (0-based).
    :param line_number: Index of the log entry to edit (0-based).
    :param new_entry: New string to replace the old log entry.
    :return: True if edit was successful, False otherwise.
    """
    logs = read_logs()
    if 0 <= line_number < len(logs):
        logs[line_number] = new_entry + '\n'  # Replace the specified line
        with open(LOG_FILE, 'w', encoding='utf-8') as f:
            f.writelines(logs)  # Write all lines back to the file
        return True
    return False  # Return False if line_number is out of range


def delete_log(line_number):
    """
    Delete a specific log entry by line number (0-based).
    :param line_number: Index of the log entry to delete (0-based).
    :return: True if deletion was successful, False otherwise.
    """
    logs = read_logs()
    if 0 <= line_number < len(logs):
        del logs[line_number]  # Remove the specified line
        with open(LOG_FILE, 'w', encoding='utf-8') as f:
            f.writelines(logs)  # Write remaining lines back to the file
        return True
    return False  # Return False if line_number is out of range


def main():
    """Menu-driven CLI for logging operations."""
    while True:
        print("\n--- Log Menu ---")
        print("1. Add log entry")
        print("2. Read all log entries")
        print("3. Edit a log entry")
        print("4. Delete a log entry")
        print("5. Exit")
        choice = input("Choose an option (1-5): ")
        if choice == '1':
            entry = input("Enter a log entry: ")
            add_log(entry)
            print("Log entry added.")
        elif choice == '2':
            logs = read_logs()
            if logs:
                print("\nAll log entries:")
                for i, log in enumerate(logs):
                    print(f"{i}: {log.strip()}")
            else:
                print("No log entries found.")
        elif choice == '3':
            logs = read_logs()
            if not logs:
                print("No log entries to edit.")
                continue
            for i, log in enumerate(logs):
                print(f"{i}: {log.strip()}")
            try:
                idx = int(input("Enter the line number to edit: "))
                new_entry = input("Enter the new log entry: ")
                if edit_log(idx, new_entry):
                    print("Log entry edited.")
                else:
                    print("Invalid line number.")
            except ValueError:
                print("Please enter a valid number.")
        elif choice == '4':
            logs = read_logs()
            if not logs:
                print("No log entries to delete.")
                continue
            for i, log in enumerate(logs):
                print(f"{i}: {log.strip()}")
            try:
                idx = int(input("Enter the line number to delete: "))
                if delete_log(idx):
                    print("Log entry deleted.")
                else:
                    print("Invalid line number.")
            except ValueError:
                print("Please enter a valid number.")
        elif choice == '5':
            print("Exiting log program.")
            break
        else:
            print("Invalid choice. Please select 1-5.")


if __name__ == "__main__":
    main()